### DISCUSSION TITLE
using oqs algorithm in existing C based application using openSSL

### DISCUSSION THREAD

**Discussion by ranjankumar23** [open-quantum-safe/oqs-provider#87] [Q&A] [answered]
Team,

I have C programmed application which is using following openSSL APIs.

SSL_METHOD *method;
SSL_CTX *ctx;
OpenSSL_add_all_algorithms();  /* Load cryptos, et.al. */
SSL_load_error_strings();   /* Bring in and register error messages */
{method = (SSL_METHOD *)TLSv1_2_server_method();}  /* Create new server-method instance */
{method = (SSL_METHOD *)TLSv1_2_client_method();}
ctx = SSL_CTX_new(method);   /* Create new context */
if ( SSL_CTX_use_certificate_file(ctx, CertFile, SSL_FILETYPE_PEM) <= 0 )
if ( SSL_CTX_use_PrivateKey_file(ctx, KeyFile, SSL_FILETYPE_PEM) <= 0 )
if ( !SSL_CTX_check_private_key(ctx) )
cert = SSL_get1_peer_certificate(ssl); /* get the server's certificate */
rx_bytes = SSL_read(ssl,cur_ptr,read_count);//tiku_new
SSL_library_init();
ssl = SSL_new(ctx);      /* create new SSL connection state */
SSL_set_fd(ssl, sockfd);
ssl_return = SSL_connect(ssl);

-----------------------Now i want this application to use OQS integrated openSSL using OQS-provider-OpenSSL.
I am thinking of following steps.
1. enable provider in openssl without recompiling openSSL 3.0.

- shall we load oqs provider and default provider in application using openSSL API 'OSSL_PROVIDER_load'?
- Or we shall set this environment variable to use local conf OPENSSL_CONF=/my/non-default/openssl.cnf?

2. Which all additional openSSL API needs to be called to use OQS algorithm for key exchange and authentication in openSSL communication?
3. Anything i am missing here?

**Comment by baentsch** [comment]
> Now i want this application to use OQS integrated openSSL using OQS-provider-OpenSSL.

One important observation: OQS-integrated TLS must use TLS1.3. It will not work for TLS1.2 (as requested above).
> I am thinking of following steps.
>
>   enable provider in openssl without recompiling openSSL 3.0.
>
>    shall we load oqs provider and default provider in application using openSSL API 'OSSL_PROVIDER_load'?
>    Or we shall set this environment variable to use local conf OPENSSL_CONF=/my/non-default/openssl.cnf?

Both options are sensible. I personally think the latter takes less code to write for yourself, while the former gives you more control (location of provider, etc). Your choice really.

>
>    Which all additional openSSL API needs to be called to use OQS algorithm for key exchange and authentication in openSSL communication?

For key exchange you need to announce the groups you're willing to utilize. See e.g., [here](https://github.com/open-quantum-safe/oqs-provider/blob/d78f950c2503456f1edca0f9ba27fa3b2397def2/test/oqs_test_groups.c#L68-L69). For auth, well, you need an OpenSSL3 where https://github.com/openssl/openssl/pull/19312 is merged. Otherwise you run into https://github.com/open-quantum-safe/oqs-provider/issues/9.

>    Anything i am missing here?

I don't think so. But actual use will show it: Test your code by connecting (hopefully successfully) against https://test.openquantumsafe.org. Be sure to select the right port for the right kem/sig combination.

**Reply by ranjankumar23** [reply]
Thanks Michael for your response.

> What is TLS API for TLS1.3 version corresponding to TLSv1_2_server_method/TLSv1_2_client_method

For response corresponding to [2] and [3] , I will try and comeback.

**Reply by baentsch** [reply]
Shouldn't that be documented at the OpenSSL website? AFAIK TLS1.3 is default by now. Edit/Add: FWIW, https://github.com/openssl/openssl/blob/master/test/helpers/ssltestlib.c is a good source for "suggested" OpenSSL API use.

**Reply by ranjankumar23** [reply]
> Shouldn't that be documented at the OpenSSL website? AFAIK TLS1.3 is default by now. Edit/Add: FWIW, https://github.com/openssl/openssl/blob/master/test/helpers/ssltestlib.c is a good source for "suggested" OpenSSL API use.

Michael, you are correct, default is highest mutually supported version by client and server. TLS_server_method(), TLS_client_method() will be sufficient.

**Reply by baentsch** [reply]
@ranjankumar23 So can we consider the question answered, then?

**Comment by ranjankumar23** [comment]
@baentsch 
Michael **_how to confirm_** that underlying communication between Two HW devices communicating using PQC enabled OpenSSL is actually using quantum safe algo for key exchange and authentication?

I saw following testing options.
1. https://test.openquantumsafe.org/
2. https://github.com/open-quantum-safe/openssl/tree/OQS-OpenSSL_1_1_1-stable#tls-demo

**Reply by baentsch** [reply]
If you direct the device(s) to connect to test.openquantumsafe.org and the relevant algorithm/test port, the device receives the "success" message when the PQ algorithm combination works (e.g., dilithium3 with kyber768 at test.openquantumsafe.org:6215):
![image](https://user-images.githubusercontent.com/57787676/210969631-eeb7b4b1-bc9d-4a2b-8b8d-d21bf0ed47a8.png)

**Reply by ranjankumar23** [reply]
Thanks Michael @baentsch for your response.

This test website can be used for verifying connectivity between my device and test server.

How to verify connectivity between my device 'A' and my device 'B'. Both devices A and B are using PQC openssl?

**Reply by baentsch** [reply]
How much control do you have over the devices? Can you simply make them output locally what they're doing? 
Blackbox testing by assuming one of your devices is a server and one is a client (and you can trust the transitivity of the tests) would allow you to verify the client with the above (as apparently done) and checking the server with a known-good client, e.g., using `oqs-curl` as stated under Note 5 on https://test.openquantumsafe.org. 
If you want to ascertain that the devices between themselves are running PQ, I'd suggest running `wireshark` to dissect the packets, e.g., using [oqs-wireshark](https://github.com/open-quantum-safe/oqs-demos/tree/main/wireshark). If you want to be 100% certain, manually re-compute the observed data exchanged -- this of course assumes you can get access to the respective devices' private keys...

**Reply by ranjankumar23** [reply]
Thanks Michael @baentsch for your detailed and clear direction.

I have full control of these devices for ssh, firmware update and debug ports.

Very pleased to know that your team has created oqs-wireshark.

**Reply by ranjankumar23** [reply]
> If you direct the device(s) to connect to test.openquantumsafe.org and the relevant algorithm/test port, the device receives the "success" message when the PQ algorithm combination works (e.g., dilithium3 with kyber768 at test.openquantumsafe.org:6215): ![image](https://user-images.githubusercontent.com/57787676/210969631-eeb7b4b1-bc9d-4a2b-8b8d-d21bf0ed47a8.png)

ranjan-PC:~/PQC-x86/.local/bin$ echo "GET /" | LD_LIBRARY_PATH=../lib ./openssl s_client -CAfile CA.crt --connect test.openquantumsafe.org:6008 --curves kyber768
CONNECTED(00000003)
depth=1 CN = oqstest_CA
verify return:1
depth=0 CN = test.openquantumsafe.org
verify return:1
---
Certificate chain
 0 s:CN = test.openquantumsafe.org
   i:CN = oqstest_CA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDRDCCASygAwIBAgIUPKsrBykp+IZOKelvOHNl2Oe09+kwDQYJKoZIhvcNAQEL
BQAwFTETMBEGA1UEAwwKb3FzdGVzdF9DQTAeFw0yMjA4MTExMjE0MjZaFw0yMzA4
MTExMjE0MjZaMCMxITAfBgNVBAMMGHRlc3Qub3BlbnF1YW50dW1zYWZlLm9yZzBZ
MBMGByqGSM49AgEGCCqGSM49AwEHA0IABPi6pyotdmVFKpuPOCyDIBHS4mGRLN6C
z+jvbWoOJCU53z4Mq/oOiOj3uhLPnu2unfh69eaTpd9jazlNxMLEprajSTBHMAsG
A1UdDwQEAwIF4DATBgNVHSUEDDAKBggrBgEFBQcDATAjBgNVHREEHDAaghh0ZXN0
Lm9wZW5xdWFudHVtc2FmZS5vcmcwDQYJKoZIhvcNAQELBQADggIBAKAX2DvzVTa6
XZCf2ZlpNiYV+yGsUsNDVRuhqH10/OVTxVhck745HijO1alyo5YVy57emYDOx+Cr
jY1O8ZZaKI9LhB4G0qXVK0JH6H9zryFN966QbwFuIv2FAD10siEAYcps+ad68np5
ehn4QYv7m4NpM6ZB5VJPQChVvC6tXqcg+GhsrY2eK1niaHomh+Zl6Osjlo54ag1h
hXT4mGX2l1LP/zqKlVM6fJHRpnHkN29QvvF8cu2ZtCY44/pFdHVAYim/JSLhf1PY
U9vXsEwHDIzAJbNzWhDBoXO9gsrGfHqUfD1R0dX0k55eBymzCsWpNXUbbeKXPRSq
1WTpRrA2FSDONSbaWz5Mxw77ztwmAcWrhH1lGVgHgzo39H2N6vM9gexWj5BrNhk3
ErCVkSCkJ8SVu0/xs2E1jhx4Yl5zI/9AeGx3fA03RDhKn3JaGm4fdVaq1d80xWUl
RgTVNhkn9ruBdh7hm2P9NN21B4zEXjXoK7FE+sBxi5QE8DL2rvCJ1TJcv6lTJp0u
gDELXJ2WjIwv/iR4UmkmBzxoQQtfG7O9KNl/hpEZ7G3t0lFq9X2lQb69YLT5krWe
OAvf7ajVvUa/5T9DmGsKNR6EAKJcxCo0H/b9XoYW2ujbI+hUiMCrJDAePU6H/b4N
2hYJWIx5UoMLPBDDt+La/APRO9JyRAcH
-----END CERTIFICATE-----
subject=CN = test.openquantumsafe.org

issuer=CN = oqstest_CA

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: ECDSA
Server Temp Key: kyber768
---
SSL handshake has read 2270 bytes and written 1606 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 256 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
DONE
ranjan@ranjan-PC:~/PQC-x86/.local/bin


@baentsch I don't see success message. Can you check whether it is successful PQC or failure.

**Reply by baentsch** [reply]
That looks OK (kyber768 server tmp key indicated). But you should be able to see the "success message", too: For reference, this is what happens when our "docker black box" is used:
```
$ docker run -it openquantumsafe/curl:0.7.2 curl --cacert /opt/oqssa/oqs-testca.pem https://test.openquantumsafe.org:6008 --curves kyber768
<!DOCTYPE html>
<html>
<head>
<title>Open Quantum Safe interop test server for quantum-safe cryptography</title>
</head>
<body>
<h1 align=center>
Successfully connected using
ecdsap256-kyber768
!
</h1>

Client-side KEM algorithm(s) indicated:
kyber768
</body>

```

--> Are you sure the HTTP GET command gets sent?

**Reply by baentsch** [reply, marked as answer]
@ranjankumar23 As your interest seems to have shifted to OpenSSL111, can we consider this question answered/closed then?

**Comment by ranjankumar23** [comment]
Yes @baentsch please close this.

**Comment by ranjankumar23** [comment]
I need help here.
https://github.com/open-quantum-safe/openssl/discussions/461#discussioncomment-6291311
